llama/model.py

```diff
@@ -190,99 +212,167 @@ def __init__(
+        head_dim = params.dim // params.n_heads
         for layer_id in range(params.n_layers):
-            self.layers.append(TransformerBlock(layer_id, params))
+            # self.layers.append(TransformerBlock(layer_id, params)) #THIS CALL IS DIFF
+            self.layers.append(TransformerBlock(layer_id, params, world_size=world_size,
+                                                rank=rank, groups=groups))
+            cache_k = torch.zeros(
+                (params.max_batch_size, params.max_seq_len, n_local_heads, head_dim)
+            )
+            cache_v = torch.zeros(
+                (params.max_batch_size, params.max_seq_len, n_local_heads, head_dim)
+            )
+            self.cache_kvs.append((cache_k, cache_v))
 
         self.norm = RMSNorm(params.dim, eps=params.norm_eps)
         self.output = ColumnParallelLinear(
-            params.dim, params.vocab_size, bias=False, init_method=lambda x: x
+            params.dim, params.vocab_size, bias=False, init_method=lambda x: x,
+            world_size=world_size, rank=rank, groups=groups
         )
 
-        self.freqs_cis = precompute_freqs_cis(
+        freqs_cis = precompute_freqs_cis(
             self.params.dim // self.params.n_heads, self.params.max_seq_len * 2
         )
+        self.register_buffer("freqs_cis", freqs_cis)
 
-    @torch.inference_mode()
-    def forward(self, tokens: torch.Tensor, start_pos: int):
+        mask = torch.full((1, 1, self.params.max_seq_len, self.params.max_seq_len),
+                          float("-inf")).to(torch.float)
+        mask = torch.triu(mask, diagonal=1)
+        self.register_buffer("mask", mask)
+
+    @torch.no_grad()
+    def forward(self, tokens: torch.Tensor, input_idexes: torch.Tensor, output_idex: torch.Tensor,
+                cache_kvs: List[Tuple[torch.Tensor, torch.Tensor]]):
         _bsz, seqlen = tokens.shape
+        assert _bsz == self.params.max_batch_size
         h = self.tok_embeddings(tokens)
-        self.freqs_cis = self.freqs_cis.to(h.device)
-        freqs_cis = self.freqs_cis[start_pos : start_pos + seqlen]
+        freqs_cis = self.freqs_cis.index_select(0, input_idexes)
 
-        mask = None
-        if seqlen > 1:
-            mask = torch.full(
-                (1, 1, seqlen, seqlen), float("-inf"), device=tokens.device
-            )
-            mask = torch.triu(mask, diagonal=start_pos + 1).type_as(h)
+        mask = self.mask.index_select(2, input_idexes)
 
-        for layer in self.layers:
-            h = layer(h, start_pos, freqs_cis, mask)
+        new_cache_kvs = []
+        for layer, cache_kv in zip(self.layers, cache_kvs):
+            h, new_cache_kv = layer(h, freqs_cis, mask, input_idexes, cache_kv)
+            new_cache_kvs.append(new_cache_kv)
         h = self.norm(h)
-        output = self.output(h).float()
-        return output
+        h = h.index_select(1, output_idex - input_idexes[0]).squeeze(dim=1)
```

Comment:
LLaMA 2 changed to output logits at all positions, so this needs to be changed accordingly.